forte/mrdsrg-spin-integrated/dsrg_mrpt2_grad/dsrg_mrpt2_deriv_tensors.cc

```diff
@@ -0,0 +1,515 @@
+/**
+ * Set CI-related and DSRG tensors.
+ */
+#include <algorithm>
+#include <map>
+#include <vector>
+#include <math.h>
+#include <numeric>
+#include <ctype.h>
+#include <string>
+
+#include "psi4/libpsi4util/PsiOutStream.h"
+#include "psi4/libpsio/psio.h"
+#include "psi4/libpsio/psio.hpp"
+#include "psi4/libqt/qt.h"
+#include "psi4/libmints/dipole.h"
+
+#include "helpers/timer.h"
+#include "ci_rdm/ci_rdms.h"
+#include "boost/format.hpp"
+#include "sci/fci_mo.h"
+#include "fci/fci_solver.h"
+#include "helpers/printing.h"
+#include "../dsrg_mrpt2.h"
+
+#include "psi4/libmints/factory.h"
+#include "psi4/libiwl/iwl.hpp"
+#include "psi4/libpsio/psio.hpp"
+
+#include "psi4/libmints/matrix.h"
+#include "psi4/libmints/wavefunction.h"
+#include "psi4/psifiles.h"
+
+#include "../master_mrdsrg.h"
+#include "helpers/timer.h"
+
+using namespace ambit;
+using namespace psi;
+
+namespace forte {
+
+/**
+ * Initializing CI integrals for the z-vector equation.
+ */
+void DSRG_MRPT2::set_ci_ints() {
+    cc = coupling_coefficients_;
+    ci = ci_vectors_[0];
+
+    cc1a_n = ambit::Tensor::build(ambit::CoreTensor, "cc1a_n", {ndets, na, na});
```

Comment:
The meanings of n and r still aren't clear to me.

Reply:
n is for "normal" and r is for "reversed". It can be identified when checked the definition, for example     
cc1a_n("Kzw") = cc.cc1a()("KJzw") * ci("J");
cc1a_r("Kzw") = cc.cc1a()("JKzw") * ci("J");
BTW, as York mentioned, this may be deprecated in the future as well when we figure out more efficient contraction routines.

Reply:
I find "n" and "r" confusing, especially because there is no apparent relation between this and the contractions of the 1RDM with the G integrals, where you also use "n" and "r". Can we label these `bra` and `ket`?

Shouldn't the coupling coefficients be symmetric? If so, `cc1a_n == cc1a_r`.

Reply:
"Smart" idea but it is wrong. When you set cc1a_n == cc1a_r you get the wrong gradient, that's one problem spotted and solved years ago.

Reply:
changed.